src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -251,6 +259,44 @@ private async Task FillRequestLists(List<Hl7.Fhir.Model.Bundle.EntryComponent> b
             }
         }
 
+        public async Task ResolveIntraBundleReferences(EntryComponent entry, Dictionary<string, (string resourceId, string resourceType)> referenceIdDictionary, CancellationToken cancellationToken)
+        {
+            IEnumerable<ResourceReference> references = entry.Resource.GetAllChildren<ResourceReference>();
+
+            foreach (ResourceReference reference in references)
+            {
+                // We've already come across this ID
+                if (referenceIdDictionary.TryGetValue(reference.Reference, out var referenceInformation))
+                {
+                    reference.Reference = $"{referenceInformation.resourceType}/{referenceInformation.resourceId}";
+                }
+                else
+                {
+                    if (reference.Reference.Contains("?", StringComparison.InvariantCulture))
+                    {
+                        string[] queries = reference.Reference.Split("?");
+                        string resourceType = queries[0];
+                        string conditionalQueries = queries[1];
+
+                        SearchResultEntry[] results = await _transactionBundleValidator.GetExistingResourceId(entry.Request.Url, resourceType, conditionalQueries, cancellationToken);
+
+                        int? resultCount = results?.Length;
+
+                        if (!resultCount.HasValue || resultCount != 1)
+                        {
+                            throw new RequestNotValidException(string.Format(Api.Resources.InvalidConditionalReference, reference.Reference));
+                        }
+
+                        string resourceId = $"{resourceType}/{results[0].Resource.ResourceId}";
+
+                        referenceIdDictionary.Add(reference.Reference, (resourceId, resourceType));
```

Comment:
For a `Patient` with id `123` this would insert a tuple of `("Patient/123", "Patient")` right? 

If it ran into the same reference again, wouldn't the first `if` statement set the references to `Patient/Patient/123`? Do we have a test case that would cover this case?

Reply:
Yes, that was the bug. I have now updated the unittest as well to cover that scenario.